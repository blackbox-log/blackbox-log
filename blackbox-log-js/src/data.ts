import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
