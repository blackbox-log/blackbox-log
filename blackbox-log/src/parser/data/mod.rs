mod event;

use alloc::vec::Vec;

pub use self::event::Event;
use super::{FrameKind, Headers, MainFrame, ParseError, ParseResult, Reader, SlowFrame};

#[derive(Debug, Clone)]
pub struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<(MainFrame, usize)>,
    // pub(crate) gps_frames: Vec<Frame>,
    // pub(crate) gps_home_frames: Vec<Frame>,
    pub(crate) slow_frames: Vec<SlowFrame>,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Stats {
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
            },
        }
    }

    pub fn parse(mut data: Reader, headers: &Headers) -> ParseResult<Self> {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        // let gps_frames = Vec::new();
        // let gps_home_frames = Vec::new();
        let mut slow_frames = Vec::new();

        slow_frames.push(headers.slow_frames.default_frame(headers));

        let mut last_kind = None;
        while let Some(kind) = data.read_u8().map(FrameKind::from_byte) {
            // TODO (rust 1.65): let-else
            let kind = if let Some(kind) = kind {
                kind
            } else {
                tracing::error!("found invalid frame byte");
                match last_kind.take() {
                    Some(FrameKind::Event) => {
                        events.pop();
                    }
                    Some(FrameKind::Intra | FrameKind::Inter) => {
                        main_frames.pop();
                    }
                    Some(FrameKind::Slow) => {
                        slow_frames.pop();
                    }
                    Some(FrameKind::Gps | FrameKind::GpsHome) | None => {}
                };

                data.skip_until_any(
                    &[
                        FrameKind::Event,
                        FrameKind::Intra,
                        FrameKind::Slow,
                        FrameKind::Gps,
                        FrameKind::GpsHome,
                    ]
                    .map(u8::from),
                );

                continue;
            };

            let result = match kind {
                FrameKind::Event => match Event::parse_into(&mut data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Intra | FrameKind::Inter => {
                    let frame = MainFrame::parse(&mut data, kind, &main_frames, headers);
                    frame.map(|frame| main_frames.push((frame, slow_frames.len() - 1)))
                }
                FrameKind::Gps => {
                    if let Some(ref gps) = headers.gps_frames {
                        gps.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::GpsHome => {
                    if let Some(ref gps_home) = headers.gps_home_frames {
                        gps_home.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS home frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::Slow => headers
                    .slow_frames
                    .parse(&mut data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(ParseError::UnexpectedEof) => {
                    tracing::warn!("found unexpected end of file");
                    break;
                }
                Err(err) => return Err(err),
            }
        }

        Ok(Self {
            events,
            main_frames,
            // gps_frames,
            // gps_home_frames,
            slow_frames,
        })
    }
}
