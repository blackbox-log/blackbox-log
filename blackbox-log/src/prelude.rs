//! Minimal set of imports for convenience.
//!
//! [`File`](`crate::File`) is not included due to avoid collisions with other
//! types.

pub use crate::frame::{Frame as _, FrameDef as _};
pub use crate::{DataParser, FieldFilter, Headers, ParserEvent};
