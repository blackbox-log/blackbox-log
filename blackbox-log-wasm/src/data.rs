use blackbox_log::prelude::*;
use blackbox_log::Reader;

use crate::{OwnedSlice, Shared, WasmFfi};

// SAFETY: field order *must* be `parser` first, then `headers`, then `data` to
// ensure correct drop order
pub struct WasmDataParser {
    parser: DataParser<'static, 'static>,
    _headers: Shared<Headers<'static>>,
    _data: Shared<OwnedSlice>,
}

impl WasmDataParser {
    pub(crate) fn new(
        headers: Shared<Headers<'static>>,
        reader: Reader<'static>,
        data: Shared<OwnedSlice>,
    ) -> Self {
        // SAFETY: this is only used to create the `DataParser`, which is guaranteed to
        // be dropped before `headers` by the declaration order in the struct
        let headers_ref = unsafe { headers.deref_static() };

        Self {
            parser: DataParser::new(reader, headers_ref),
            _headers: headers,
            _data: data,
        }
    }

    fn main_frame_count(&self) -> usize {
        self.parser.stats().counts.main
    }

    fn gps_frame_count(&self) -> usize {
        self.parser.stats().counts.gps
    }
}

impl WasmFfi for WasmDataParser {}

#[no_mangle]
unsafe extern "wasm" fn data_free(ptr: *mut WasmDataParser) {
    let parser = WasmDataParser::from_wasm(ptr);
    drop(parser);
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_mainFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.main_frame_count();
    parser.into_wasm();
    count
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_gpsFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.gps_frame_count();
    parser.into_wasm();
    count
}
