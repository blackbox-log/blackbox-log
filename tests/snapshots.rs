use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
