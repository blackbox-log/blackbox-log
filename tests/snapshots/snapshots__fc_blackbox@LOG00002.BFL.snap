---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/fc-blackbox/src/test-data/LOG00002.BFL"
---
File(
  count: 1,
  logs: [
    Ok(Log(
      headers: Headers(
        firmware_revision: "Betaflight 4.2.6 (a4b6db1e7) STM32F405",
        firmware: Betaflight("4.2.6"),
        firmware_date: Some(Err("Jan  5 2021 19:07:23")),
        board_info: Some("AIRB OMNIBUSF4"),
        craft_name: None,
        debug_mode: GyroFiltered,
        disabled_fields: [],
        features: [
          RxSerial,
          AirMode,
          AntiGravity,
        ],
        unknown: {
          "dyn_notch_width_percent": "0",
          "rc_smoothing_active_cutoffs": "0, 90",
          "acc_hardware": "0",
          "deadband": "0",
          "tpa_rate": "65",
          "rc_smoothing_cutoffs": "0, 0",
          "gyro_lowpass2_type": "0",
          "rate_limits": "1998,1998,1998",
          "dterm_notch_cutoff": "0",
          "pid_process_denom": "1",
          "gyro_cal_on_first_arm": "0",
          "yaw_lowpass_hz": "0",
          "acc_limit_yaw": "0",
          "dyn_notch_max_hz": "400",
          "yaw_deadband": "0",
          "mag_hardware": "1",
          "maxthrottle": "2000",
          "vbat_pid_gain": "0",
          "I interval": "256",
          "dterm_lowpass_dyn_hz": "70,170",
          "use_integrated_yaw": "0",
          "dshot_bidir": "0",
          "ff_boost": "15",
          "gyro_sync_denom": "1",
          "rpm_notch_lpf": "150",
          "rc_smoothing_filter_type": "1, 1",
          "serialrx_provider": "2",
          "dshot_idle_value": "550",
          "feedforward_weight": "90,95,90",
          "d_min_advance": "20",
          "rc_smoothing_debug_axis": "0",
          "gyro_rpm_notch_q": "500",
          "rc_interpolation_channels": "2",
          "dterm_filter2_type": "0",
          "ff_interpolate_sp": "2",
          "gyro_lowpass_type": "0",
          "Log start datetime": "0000-01-01T00:00:00.000+00:00",
          "dterm_filter_type": "0",
          "vbatcellvoltage": "330,350,430",
          "anti_gravity_mode": "0",
          "rc_interpolation_interval": "19",
          "rc_smoothing_type": "1",
          "iterm_relax": "1",
          "gyro_rpm_notch_harmonics": "2",
          "rc_interpolation": "2",
          "tpa_breakpoint": "1350",
          "d_min_gain": "37",
          "dterm_lowpass_hz": "150",
          "rates": "70,70,70",
          "iterm_relax_cutoff": "15",
          "d_min": "23,25,0",
          "acc_limit": "0",
          "ff_spike_limit": "60",
          "ff_max_rate_limit": "100",
          "gyro_lowpass_hz": "200",
          "gyro_notch_hz": "0,0",
          "anti_gravity_threshold": "250",
          "acc_lpf_hz": "1000",
          "baro_hardware": "1",
          "motor_pwm_protocol": "7",
          "rc_rates": "100,100,100",
          "pidAtMinThrottle": "1",
          "pitchPID": "46,90,38",
          "magPID": "40",
          "P interval": "8",
          "pidsum_limit_yaw": "400",
          "pidsum_limit": "500",
          "rates_type": "0",
          "dyn_notch_min_hz": "80",
          "iterm_relax_type": "1",
          "dterm_rpm_notch_min": "100",
          "gyro_lowpass_dyn_hz": "200,500",
          "motor_pwm_rate": "480",
          "thr_mid": "50",
          "dyn_notch_q": "250",
          "dterm_notch_hz": "0",
          "feedforward_transition": "0",
          "gyro_rpm_notch_min": "80",
          "rollPID": "42,85,35",
          "looptime": "125",
          "thr_expo": "0",
          "yawPID": "45,90,0",
          "levelPID": "50,50,75",
          "rc_expo": "0,0,0",
          "dterm_rpm_notch_harmonics": "0",
          "airmode_activate_throttle": "25",
          "P ratio": "32",
          "anti_gravity_gain": "3500",
          "gyro_hardware_lpf": "0",
          "dterm_lowpass2_hz": "150",
          "use_unsynced_pwm": "0",
          "rc_smoothing_rx_average": "0",
          "dterm_rpm_notch_q": "500",
          "iterm_windup": "100",
          "abs_control_gain": "0",
          "gyro_notch_cutoff": "0,0",
          "rc_smoothing_auto_factor": "10",
          "gyro_lowpass2_hz": "250",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 0,
          main: 199975,
          slow: 25,
          gps: 0,
          gps_home: 0,
        ),
      ),
      capped: true,
      main: MainFrame(
        count: 199975,
        last_iteration: 1599750,
        time: FieldSnapshot(
          min: 0,
          max: 206899760,
          unique: 199975,
          histogram: (12489, 12409, 12484, 12724, 12080, 12958, 12731, 12337, 12468, 12231, 12358, 12813, 11953, 12824, 12809, 12307),
        ),
        fields: Fields([
          Field(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -168,
              max: 174,
              unique: 339,
              histogram: (34158, 22167, 18666, 15670, 13416, 12647, 11732, 10940, 9726, 8972, 8355, 7732, 7014, 6536, 6141, 6103),
            ),
          ),
          Field(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -57,
              max: 140,
              unique: 183,
              histogram: (30316, 19700, 18151, 16693, 15145, 13566, 12336, 11078, 10416, 9315, 8666, 7883, 7441, 6941, 6446, 5882),
            ),
          ),
          Field(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -199,
              max: 281,
              unique: 477,
              histogram: (36194, 20074, 14404, 12779, 12165, 11537, 10585, 9854, 9518, 9536, 9313, 9177, 9020, 8575, 8858, 8386),
            ),
          ),
          Field(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -100,
              max: 53,
              unique: 96,
              histogram: (52617, 37702, 29904, 22832, 16444, 11740, 8351, 5894, 4177, 2979, 2168, 1661, 1187, 1008, 735, 576),
            ),
          ),
          Field(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -58,
              max: 48,
              unique: 92,
              histogram: (50149, 36233, 29530, 23103, 17227, 12698, 9061, 6458, 4721, 3351, 2340, 1788, 1251, 926, 623, 516),
            ),
          ),
          Field(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1000,
              unique: 1,
            ),
          ),
          Field(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "vbatLatest",
            unit: Voltage,
            history: FieldSnapshot(
              min: 0,
              max: 70,
              unique: 8,
              histogram: (34324, 0, 527, 0, 46796, 0, 79, 0, 7334, 0, 99857, 0, 59, 0, 10999, 0),
            ),
          ),
          Field(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -128,
              max: 125,
              unique: 252,
              histogram: (26937, 27764, 21819, 17414, 15103, 13795, 12340, 10859, 9794, 8589, 7668, 7054, 6327, 5717, 4676, 4119),
            ),
          ),
          Field(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -95,
              max: 40,
              unique: 127,
              histogram: (25238, 26454, 22671, 19141, 16106, 13821, 12343, 11041, 9905, 9068, 8045, 6907, 5926, 5102, 4338, 3869),
            ),
          ),
          Field(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -195,
              max: 139,
              unique: 335,
              histogram: (31266, 26661, 16279, 14238, 13642, 11959, 11553, 10850, 10160, 9825, 9006, 8064, 6817, 6575, 6749, 6331),
            ),
          ),
          Field(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -192,
              max: 1074,
              unique: 1259,
              histogram: (11992, 11843, 12077, 12302, 12278, 11926, 11609, 11351, 11970, 11740, 12085, 12353, 11983, 14007, 16524, 13935),
            ),
          ),
          Field(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -774,
              max: 700,
              unique: 1467,
              histogram: (11669, 12256, 12903, 12347, 12327, 12251, 12495, 18567, 13543, 11916, 11394, 12619, 11632, 11581, 11232, 11243),
            ),
          ),
          Field(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: 0,
              max: 1956,
              unique: 1459,
              histogram: (12701, 13453, 13147, 13018, 12934, 12796, 12791, 12472, 12310, 11821, 11996, 11902, 12155, 12043, 12128, 12308),
            ),
          ),
          Field(
            name: "debug[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -128,
              max: 125,
              unique: 252,
              histogram: (26937, 27764, 21819, 17414, 15103, 13795, 12340, 10859, 9794, 8589, 7668, 7054, 6327, 5717, 4676, 4119),
            ),
          ),
          Field(
            name: "debug[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -95,
              max: 40,
              unique: 127,
              histogram: (25238, 26454, 22671, 19141, 16106, 13821, 12343, 11041, 9905, 9068, 8045, 6907, 5926, 5102, 4338, 3869),
            ),
          ),
          Field(
            name: "debug[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -195,
              max: 139,
              unique: 335,
              histogram: (31266, 26661, 16279, 14238, 13642, 11959, 11553, 10850, 10160, 9825, 9006, 8064, 6817, 6575, 6749, 6331),
            ),
          ),
          Field(
            name: "debug[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          Field(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 158,
              unique: 1,
            ),
          ),
          Field(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 158,
              unique: 1,
            ),
          ),
          Field(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 158,
              unique: 1,
            ),
          ),
          Field(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 158,
              unique: 1,
            ),
          ),
        ]),
      ),
      slow: SlowFrame(
        count: 25,
        fields: Fields([
          Field(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {},
          ),
          Field(
            name: "stateFlags",
            unit: State,
            history: {},
          ),
          Field(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 25,
            },
          ),
          Field(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 0,
              no: 25,
            ),
          ),
          Field(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 0,
              no: 25,
            ),
          ),
        ]),
      ),
    )),
  ],
)
